File: OC/MVC/tableView的header下拉放大/tableView的header下拉放大/ViewController.m
//
//  ViewController.m
//  tableView的header下拉放大
//
//  Created by mic on 2017/8/28.
//  Copyright © 2017年 yunfu. All rights reserved.
//

#import "ViewController.h"

#define KScreenWidth  [UIScreen mainScreen].bounds.size.width
#define KScreenHeight [UIScreen mainScreen].bounds.size.height

#define HeaderHeight 200

@interface ViewController ()<UITableViewDelegate, UITableViewDataSource>

@property (nonatomic, strong) UITableView *tableView;
@property(nonatomic, strong) UIImageView *bgImgView;    //背景图片

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor whiteColor];
    
    [self.view addSubview:self.tableView];
}

/**
 *tableView的懒加载
 */
- (UITableView *)tableView{
    if(!_tableView){
        _tableView = [[UITableView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, self.view.frame.size.height)];
        
        _tableView.delegate = self;
        _tableView.dataSource = self;
        
        _tableView.tableFooterView = [UIView new];
        _tableView.showsVerticalScrollIndicator = NO;
        _tableView.showsHorizontalScrollIndicator = NO;
        
        UIView * headerView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, KScreenWidth, HeaderHeight)];
        headerView.backgroundColor = [UIColor clearColor];
        
        UIImageView * image = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, KScreenWidth, HeaderHeight)];
        image.image = [UIImage imageNamed:@"b"];
        [headerView addSubview:image];
        self.bgImgView = image;
        
        self.tableView.tableHeaderView = headerView;
    }
    return _tableView;
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return 20;
}

#pragma mark - cellForRowAtIndexPath
-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    UITableViewCell * cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
    
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"cell"];
    }
    
    cell.textLabel.text = [NSString stringWithFormat:@"这是第%ld个cell",indexPath.row];
    
    return cell;
}

- (void)scrollViewDidScroll:(UIScrollView *)scrollView{
    CGFloat offsetY = scrollView.contentOffset.y;
    
    if (offsetY < 0) {
        CGFloat totalOffsetY = HeaderHeight + ABS(offsetY);
        CGFloat f = totalOffsetY / HeaderHeight;
        
        self.bgImgView.frame = CGRectMake(- (KScreenWidth * f - KScreenWidth) / 2, offsetY, KScreenWidth * f, totalOffsetY);
    }
}

@end

File: OC/MVC/代码/LSTableViewHeaderView--Swift/LSTableViewHeaderView--Swift/ViewController.swift
//
//  ViewController.swift
//  LSTableViewHeaderView--Swift
//
//  Created by lisonglin on 27/06/2017.
//  Copyright © 2017 lisonglin. All rights reserved.
//

import UIKit

let KScreenWidth: CGFloat = UIScreen.main.bounds.size.width
let KScreenHeight: CGFloat = UIScreen.main.bounds.size.height
let HeaderHeight: CGFloat = 200.0


class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
    
    var tableView: UITableView?
    var bgImage : UIImageView?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        self.view.backgroundColor = UIColor.white
        
        self.initTableView()
    }
    
    func initTableView() -> Void {
        self.tableView = UITableView.init(frame: CGRect.init(x: 0, y: 0, width: KScreenWidth, height: KScreenHeight), style: UITableViewStyle.plain)
        self.tableView?.backgroundColor = UIColor.clear
        
        self.tableView?.delegate = self
        self.tableView?.dataSource = self
        
        
        let headerView = UIView.init(frame: CGRect.init(x: 0, y: 0, width: KScreenWidth, height: HeaderHeight))
        
        headerView.backgroundColor = UIColor.white
        
        let image = UIImageView.init(frame: CGRect.init(x: 0.0, y: 0.0, width: KScreenWidth, height: HeaderHeight))
        
        image.image = UIImage.init(named: "b")
        
        headerView.addSubview(image)
        
        self.bgImage = image
        
        self.tableView?.tableHeaderView = headerView
        
        self.view .addSubview(self.tableView!)
        
    }
    
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 20
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        var cell = tableView.dequeueReusableCell(withIdentifier: "cell")
        
        if (cell == nil) {
            cell = UITableViewCell.init(style: UITableViewCellStyle.default, reuseIdentifier: "cell")
        }
        
        cell?.textLabel?.text =  "这是第"+"\(indexPath.row)"+"个cell"
        
        return cell!

    }
    
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        
        let offsetY: CGFloat = CGFloat(scrollView.contentOffset.y)
        
        if offsetY < 0 {
            let totalOffsetY = HeaderHeight + abs(offsetY)
            
            let f = CGFloat(totalOffsetY / HeaderHeight)
            
            self.bgImage!.frame = CGRect.init(x: -(KScreenWidth * f - KScreenWidth) / 2, y: offsetY, width: KScreenWidth * f, height: totalOffsetY)
            
        }
        
    }

}

